### PR TITLE
separate dockerimage for 4.5 and 5.0+

### PR DESCRIPTION
**Pull request by glicht**
## Status
In Progress

## Description
Infra support for specifying different docker images for 4.5 and 5.0. Allows for easily upgrading docker images where there is no code change.

Basic idea is add to the integration/script yml a new field: `dockerimage45`. When this field is set, the package creator will create 2 unified yml files. One for `toversion: 4.5.9` and one `fromversion: 5.0.0`.

## Does it break backward compatibility?
   - No
   
## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
* Includes as a first integration for the first flow: kafka v2.
* Break build if the generated integration file exists. As can be seen here: https://circleci.com/gh/demisto/content/32943?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
* Remove old Exchange2016_Compliance.yml as a result of above check

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](link)
- [x] [CHANGELOG](https://github.com/demisto/content/blob/docker-45-creator/Integrations/Kafka_V2/CHANGELOG.md)

### CODE REVIEW

**Comment by glicht** [comment]
@yaakovi please checkout locally this branch and run pre-commit hook

**Comment by glicht** [comment]
@yaakovi I've changed the validation to work against `origin/master` instead of `master`. circleci doesn't setup master when working on a branch.

File: package_creator.py [hunk 1]
```diff
@@ -70,9 +73,34 @@ def merge_script_package_to_yml(package_path, dir_name, dest_path=""):
...
+    if 'dockerimage45' in script_obj:
+        # we need to split into two files 45 and 50. Current one will be from version 5.0
+        yml_text = re.sub(r'^\s*dockerimage45:.*\n?', '', yml_text, flags=re.MULTILINE)  # remove the dockerimage45 line
+        yml_text45 = yml_text
+        if 'fromversion' in yml_data:
+            yml_text = re.sub(r'^fromversion:.*$', 'fromversion: 5.0.0', yml_text, flags=re.MULTILINE)
+        else:
+            yml_text = 'fromversion: 5.0.0\n' + yml_text
+        if 'toversion' in yml_data:
+            yml_text45 = re.sub(r'^toversion:.*$', 'toversion: 4.5.9', yml_text, flags=re.MULTILINE)
```
Comment (yaakovi) [inline package_creator.py]:
```suggestion
            yml_text45 = re.sub(r'^toversion:.*$', 'toversion: 4.5.9', yml_text45, flags=re.MULTILINE)
```

File: package_creator.py [hunk 2]
```diff
@@ -70,9 +73,34 @@ def merge_script_package_to_yml(package_path, dir_name, dest_path=""):
...
+        else:
+            yml_text45 = 'toversion: 4.5.9\n' + yml_text45
+        yml_text45 = re.sub(r'(^\s*dockerimage:).*$', r'\1 ' + script_obj.get('dockerimage45'), yml_text45, flags=re.MULTILINE)
+        output_path45 = re.sub(r'\.yml$', '_45.yml', output_path)
+        output_map = {
+            output_path: yml_text,
+            output_path45: yml_text45
+        }
+
+    for k, v in output_map.items():
```
Comment (yaakovi) [inline package_creator.py]: instead of `k, v` use `file_path, file_text` (easier to read)

**Comment by yaakovi** [comment]
added a small workaround (since `validate_files.py` is being refactored soon) to validate only the content entities.

as part of this PR (https://github.com/demisto/content/pull/4180) the `modified_files` includes the schema files and not only content entities.

**Comment by Itay4** [comment]
can we apply this change on https://github.com/demisto/content/tree/master/Scripts/RegexGroups in this pr ?

**Comment by glicht** [comment]
@yaakovi updates to content_creator to support dockerimage45:

* removed bundle_pre (3.5)
* bundle_test contains only test stuff
* output of merge is straight to the bundle_post dir
* support dockerimage45 for scripts/integrations which are not package format

**Comment by glicht** [comment]
@Itay4 you can add to the PR the regexgroup. not sure what needs to be changed there.

**Comment by glicht** [comment]
@Itay4 added regexgroup. please take a look

File: package_creator.py [hunk 3]
```diff
@@ -23,6 +25,53 @@
...
+    Returns:
+        dict -- dictionary mapping output path to text data
+    """
+    output_map = {output_path: yml_text}
+    if 'dockerimage45' in script_obj:
+        # we need to split into two files 45 and 50. Current one will be from version 5.0
+        yml_text = re.sub(r'^\s*dockerimage45:.*\n?', '', yml_text, flags=re.MULTILINE)  # remove the dockerimage45 line
+        yml_text45 = yml_text
+        if 'fromversion' in yml_data:
+            yml_text = re.sub(r'^fromversion:.*$', 'fromversion: 5.0.0', yml_text, flags=re.MULTILINE)
```
Comment (anara123) [inline package_creator.py]: why you replace fromversion with 5.0.0? the same toversion with4.5.9?
what if toversion was 4.0.0?
the same for fromversion in the future could be 5.5.0

Comment (glicht) [reply package_creator.py]: You are correct. I assumed the dockerimage45 will be used only when needed. Will add a validation.

File: content_creator.py [hunk 4]
```diff
@@ -63,44 +62,48 @@ def convert_incident_fields_to_array():
...
-            print ' - marked as post: %s (%s)' % (ver, path, )
-            shutil.copyfile(path, os.path.join(bundle_post, os.path.basename(path)))
-            shutil.copyfile(path, os.path.join(bundle_test, os.path.basename(path)))
-            post_files += 1
-        else:
-            # add the file to both bundles
-            print ' - marked as pre: %s (%s)' % (ver, path, )
-            shutil.copyfile(path, os.path.join(bundle_pre, os.path.basename(path)))
-            shutil.copyfile(path, os.path.join(bundle_post, os.path.basename(path)))
+        print ' - processing: %s (%s)' % (ver, path, )
```
Comment (Itay4) [inline content_creator.py]: better to use format

File: content_creator.py [hunk 5]
```diff
@@ -63,44 +62,48 @@ def convert_incident_fields_to_array():
...
+        copy_yaml_post(path, os.path.join(bundle_post, os.path.basename(path)), yml_info)
+        post_files += 1
+    print ' - total files: %d' % (post_files, )
```
Comment (Itay4) [inline content_creator.py]: better to use format

File: docs/docker/README.MD [hunk 6]
```diff
@@ -4,7 +4,17 @@
...
 
 # Why Use Docker?
-Primarily we use docker to run python scripts in a controlled environment. They run isolated from the server to prevent someone from accidentally damaging the server. By packaging libraries and dependencies together, we can prevent unknown issues from occurring since the environment is all the same.
+Primarily we use docker to run python scripts and integrations in a controlled environment. They run isolated from the server to prevent someone from accidentally damaging the server. By packaging libraries and dependencies together, we can prevent unknown issues from occurring since the environment is all the same.
+
+## Script/Integration Configuration
+Specifying which docker image to use is done in the Demisto IDE (Open: Settings -> Docker image name). If you don't specify a docker image, a default docker image using Python 2.7 is used. For new scripts and integrations, unless there is a specific reason not to use Python 2 (for example: a need to use a library which is not available for Python 3), we require using a Python 3 image.
```
Comment (Itay4) [inline docs/docker/README.MD]: worth mentioning here that it can also be done in the pycharm plugin

File: package_creator.py [hunk 7]
```diff
@@ -23,6 +26,59 @@
 IMAGE_PREFIX = 'data:image/png;base64,'
 
 
+def write_yaml_with_docker(output_path, yml_text, yml_data, script_obj):
```
Comment (Itay4) [inline package_creator.py]: can we add test for that one?

**Comment by kirbles19** [comment]
@glicht Done.

File: content_creator.py (hunk 4 above)
Comment (glicht) [reply content_creator.py]: This whole script needs to be moved to py3 and cleaned up. We will do this as a separate initiative.

File: docs/docker/README.MD (hunk 6 above)
Comment (glicht) [reply docs/docker/README.MD]: I think it's ok. I want to keep things short.

File: package_creator.py (hunk 7 above)
Comment (glicht) [reply package_creator.py]: We don't have any tests for this script. We need to refactor it. Let's do this as part of demisto-sdk.

**Pull request merged**